### PR TITLE
feat: assigned bookings per period

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

Fixes: CAL-4766 #17763 

Loom: https://www.loom.com/share/c0b6994f72c64d9cbc11bbd25e56b554

## How should this be tested?

Change seed size of your acme organization to 100 or higher in your seed script
Test the table is working as expected

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/features/data-table/components/DataTableSkeleton.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { TableNew, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@calcom/ui";
+
+interface DataTableSkeletonProps {
+  columns: number;
+  rows?: number;
+  columnWidths?: number[];
+}
+
+export function DataTableSkeleton({ columns, rows = 10, columnWidths = [] }: DataTableSkeletonProps) {
```
Comment (sean-brydon) [inline packages/features/data-table/components/DataTableSkeleton.tsx:9]: Extracted a skeleton version of our data-table

File: packages/features/insights/server/trpc-router.ts
Comment (sean-brydon) [inline packages/features/insights/server/trpc-router.ts:1]: Standard TRPC handlers calling the insights class

File: packages/prisma/seed-utils.ts [hunk 2]
```diff
@@ -610,11 +610,18 @@ export async function seedRoutingFormResponses(
     const shuffledOptions = [...attributeRaw[2].options].sort(() => Math.random() - 0.5);
     const selectedSkills = shuffledOptions.slice(0, numSkills);
 
+    // Generate a random date within the last 30 days
```
Comment (sean-brydon) [inline packages/prisma/seed-utils.ts:613]: Make sure the booking and form response line up

File: packages/features/insights/server/routing-events.ts [hunk 3]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+    const teamConditions = [];
+
+    // @ts-expect-error it does exist but TS isn't smart enough when it's number or int filter
+    if (formsWhereCondition.teamId?.in) {
+      // @ts-expect-error same as above
+      teamConditions.push(`f."teamId" IN (${formsWhereCondition.teamId.in.join(",")})`);
+    }
+    if (routingFormId) {
+      teamConditions.push(`f.id = '${routingFormId}'`);
+    }
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:697]: Build the conditions to make sure to bring back the selected TeamIDs and routing forms in raw sql.

File: packages/features/insights/server/routing-events.ts [hunk 4]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+
+    const searchCondition = searchQuery
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:699]: IF a search query is present check it is case insensitive

File: packages/features/insights/server/routing-events.ts [hunk 5]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+          u."avatarUrl"
+        FROM "App_RoutingForms_FormResponse" r
+        JOIN "App_RoutingForms_Form" f ON r."formId" = f.id
+        JOIN "Booking" b ON r."routedToBookingUid" = b.uid
+        JOIN "users" u ON b."userId" = u.id
+        WHERE r."routedToBookingUid" IS NOT NULL
+        AND r."createdAt" >= ${startDate}
+        AND r."createdAt" <= ${endDate}
+        ${searchCondition}
+        ${whereClause}
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:731]: Search input + where clause from routingFormId and team IDs

File: packages/features/insights/server/routing-events.ts [hunk 6]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+        SELECT date + (CASE 
+          WHEN ${dayjsPeriod} = 'day' THEN interval '1 day'
+          WHEN ${dayjsPeriod} = 'week' THEN interval '1 week'
+          WHEN ${dayjsPeriod} = 'month' THEN interval '1 month'
+        END)
+        FROM date_range
+        WHERE date < date_trunc(${dayjsPeriod}, ${endDate}::timestamp + interval '1 day')
+      ),
+      -- Get all unique users we want to show
+      all_users AS (
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:766]: CTE to select by userIDs that we generate in the usersQuery above.

File: packages/features/insights/server/routing-events.ts [hunk 7]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+      -- Get periods with pagination
+      paginated_periods AS (
+        SELECT date as period_start
+        FROM date_range
+        ORDER BY date ASC
+        ${cursor ? Prisma.sql`OFFSET ${parseInt(cursor, 10)}` : Prisma.empty}
+        LIMIT ${limit}
+      ),
+      -- Generate combinations for paginated periods
+      date_user_combinations AS (
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:778]: Generate a userId -> period map

File: packages/features/insights/server/routing-events.ts [hunk 8]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+          stat.total_bookings > average
+            ? "above_average"
+            : stat.total_bookings === median
+            ? "median"
+            : stat.total_bookings < average
+            ? "below_average"
+            : "at_average",
+      };
+      return acc;
+    }, {} as Record<number, { total: number; performance: "above_average" | "at_average" | "below_average" | "median" }>);
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:868]: I couldnt do figure out how to do this in SQL so we calculate the avergae for the period per user here and add it in to each row

File: packages/features/insights/server/routing-events.ts [hunk 9]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+        data: users.map((user) => ({
+          ...user,
+          performance: userPerformance[user.id]?.performance || "no_data",
+          totalBookings: userPerformance[user.id]?.total || 0,
+        })),
+        nextCursor: hasMoreUsers ? users[users.length - 1].id : undefined,
+      },
+      periodStats: {
+        data: periodStats,
+        nextCursor: hasMorePeriods ? (currentPeriodOffset + limit).toString() : undefined,
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:885]: Users + their period data paginated in X and Y

File: packages/features/insights/server/routing-events.ts [hunk 10]
```diff
@@ -635,6 +645,483 @@ class RoutingEventsInsights {
...
+      },
+    };
+  }
+
+  static async routedToPerPeriodCsv({
```
Comment (sean-brydon) [inline packages/features/insights/server/routing-events.ts:890]: The same as above but we don't use pagination we load all data for the current period and filters selected

Comment (eunjae-lee) [reply packages/features/insights/server/routing-events.ts:890]: I'm a bit concerned about how to keep these two in sync 🥲

**Review by eunjae-lee — COMMENTED** [review]
You said you need to do more UI work, so I'm not approving it. I've looked closely at the queries , and they seem okay.

File: packages/features/data-table/components/DataTableSkeleton.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import { TableNew, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@calcom/ui";
+
+interface DataTableSkeletonProps {
+  columns: number;
+  rows?: number;
+  columnWidths?: number[];
```
[the remark this answers is not in the document]
Comment (sean-brydon) [reply packages/features/data-table/components/DataTableSkeleton.tsx:6]: This is a loading state so we don't really have access to this - especially in this case where a lot of the headers are dynamic. 

Great tip for the other tables tho!

**Review by eunjae-lee — APPROVED** [review]
It's working well locally.

Before I seeded a big amount of data, it looks a bit empty.

Screenshot 2024-12-10 at 10 32 57

but it's much better once some data is there.

And I guess you've removed the arrow icon on the top right corner? I don't see it.

Screenshot 2024-12-10 at 10 33 35

Huge work!

File: packages/features/insights/components/RoutedToPerPeriod.tsx [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+                    );
+                  })}
+                </TableRow>
+              </TableHeader>
+              <TableBody className="relative">
+                {processedData.map((row, index) => {
+                  return (
+                    <TableRow
+                      key={row.id}
+                      ref={index === processedData.length - 1 ? loadMoreRef : undefined}
```
[the remark this answers is not in the document]
Comment (eunjae-lee) [reply packages/features/insights/components/RoutedToPerPeriod.tsx:408]: If I understand this correctly, if we send the same `loadMoreRef` to all the `TableRow`s, then each `TableRow` will call `loadMoreRef` with their own ref, meaning `useInViewObserver` will replace `node` every time it renders a row.

Is it correct? If that's the case, I don't know if it's good to use `useInViewObserver` in table rows, compared to `useFetchMoreOnBottomReached`, which observes only the container.

[the remark this answers is not in the document]
Comment (sean-brydon) [reply packages/features/insights/components/RoutedToPerPeriod.tsx:408]: Maybe youre right - will tackle this in a follow up. As we really only care about the container here

**Pull request merged**